modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheIgniteOutOfMemoryExceptionTest.java

```diff
@@ -200,10 +251,250 @@ public void testHugeEntry() throws Exception {
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testContainsKeyTwoNodes() throws Exception {
+        CacheConfiguration<Integer, Object> ccfg = cacheConfiguration(ATOMIC, HUGE_ATOMIC_CACHE_NAME, HUGE_DATA_REGION_NAME)
+            .setReadFromBackup(false)
+            .setBackups(1);
+
+        IgniteCache<Integer, Object> cache = grid(0).getOrCreateCache(ccfg);
+
+        testContainsFromBackup0(cache, 1);
+    }
+
+    /**
+     * Tests that contains operation does not require loading the whole entry.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testContainsKeysTwoNodes() throws Exception {
+        CacheConfiguration<Integer, Object> ccfg = cacheConfiguration(ATOMIC, HUGE_ATOMIC_CACHE_NAME, HUGE_DATA_REGION_NAME)
+            .setReadFromBackup(false)
+            .setBackups(1);
+
+        IgniteCache<Integer, Object> cache = grid(0).getOrCreateCache(ccfg);
+
+        testContainsFromBackup0(cache, 3);
+    }
+
+    /**
+     * Tests that contains operation does not require loading the whole entry.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testContainsKeyWithExpiryPolicyTwoNodes() throws Exception {
+        CacheConfiguration<Integer, Object> ccfg = cacheConfiguration(ATOMIC, HUGE_ATOMIC_CACHE_NAME, HUGE_DATA_REGION_NAME)
+            .setReadFromBackup(false)
+            .setBackups(1)
+            .setEagerTtl(false)
+            .setExpiryPolicyFactory(CreatedExpiryPolicy.factoryOf(new Duration(TimeUnit.MINUTES, 1)));
+
+        IgniteCache<Integer, Object> cache = grid(0).getOrCreateCache(ccfg);
+
+        testContainsFromBackup0(cache, 1);
+    }
+
+    public void testContainsFromBackup0(IgniteCache<Integer, Object> cache, int cnt) throws Exception {
+        startGrid(1);
+
+        try {
+            awaitPartitionMapExchange();
+
+            testContains(cache, backupKeys(cache, 1, 0), true);
+        }
+        finally {
+            stopGrid(1);
+        }
+    }
+
+    /**
+     * Tests that contains operation does not require loading the whole entry.
+     *
+     * @param cache Cache.
+     * @param keys Keys.
+     * @param expectContains Expected result.
+     */
+    private void testContains(IgniteCache<Integer, Object> cache, Collection<Integer> keys, boolean expectContains) {
+        Runtime.getRuntime().gc();
+
+        int blobSize = (int) (512 / keys.size() * U.MB);
+
+        for (Integer k : keys)
+            cache.put(k, new byte[blobSize]);
+
+        long usedMemory = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
+        long totalFreeMemory = Runtime.getRuntime().maxMemory() - usedMemory;
+
+        // Let's occupy all free memory.
+        List<Object> unused = new ArrayList<>();
+        IgniteBiTuple<HotSpotDiagnosticMXBean, VMOption> mbean = disableHeapDumpOnOutOfMemoryError();
+        try {
+            while (blobSize < totalFreeMemory) {
```

Comment:
This condition doesn't include possible GC

Reply:
This loop was reworked.